### PR TITLE
fix(aws-eks): Support for http proxy in EKS onEvent lambda

### PR DESCRIPTION
**Pull request by ryparker**
## Summary

Currently when a user wants to route all of the EKS lambda's SDK requests through a proxy then they are [instructed to configure an env var named `HTTP_PROXY` or `http_proxy`](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-eks-readme.html#cluster-handler).

e.g.
```ts
const cluster = new eks.Cluster(this, 'hello-eks', {
  version: eks.KubernetesVersion.V1_21,
  clusterHandlerEnvironment: {
    'http_proxy': 'http://proxy.myproxy.com'
  }
});
```

However the JS SDK [requires further configuration to enable proxy support](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/node-configuring-proxies.html).

This PR:
- Adds a `package.json` with the dependency 'proxy-agent' to the `cluster-resource-handler/` lambda bundle
- Uses `NodeJSFunction` to install lambda dependencies and bundle.
- Adds a condition that checks the environment for `HTTP_PROXY` or `http_proxy` values. If present then configures the aws-sdk to use that proxy (using `proxy-agent`).

Note: I placed the `proxy-agent` in the `devDependencies` of `package.json`. If the dependency is placed in the `dependencies` section then the CDK builder [throws an error: `NPM Package cluster-resources-handler inside jsii package '@aws-cdk/aws-eks', can only have devDependencies`](https://github.com/aws/aws-cdk/blob/7dae114b7aac46321b8d8572e6837428b4c633b2/tools/pkglint/lib/rules.ts#L1332)

Fixes: SIM D29159517, https://github.com/aws/aws-cdk/issues/12469

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-provider.ts [hunk 1]
```diff
@@ -58,8 +59,8 @@ export class ClusterResourceProvider extends NestedStack {
   private constructor(scope: Construct, id: string, props: ClusterResourceProviderProps) {
     super(scope as CoreConstruct, id);
 
-    const onEvent = new lambda.Function(this, 'OnEventHandler', {
-      code: lambda.Code.fromAsset(HANDLER_DIR),
+    const onEvent = new NodejsFunction(this, 'OnEventHandler', {
```
Comment (ryparker) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-provider.ts:62]: Using `NodejsFunction` so that NPM dependencies are installed at synth time.

**Comment by ryparker** [comment]
> ### AWS CodeBuild CI Report
> * CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
> * Commit ID: [483ad37](https://github.com/aws/aws-cdk/commit/483ad378df383115d48af06a0fff3cf7ded12622)
> * Result: FAILED
> * [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=82cc2184-497b-4588-939b-7ee51b658641%2Fbuild.log) (available for 30 days)
> 
> _Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)_

Fix for tests is on the way. Deploying and validating integration test changes takes a while with EKS.

**Review by otaviomacedo — COMMENTED** [review]
Looks good to me, but the integration tests of aws-stepfunctions-tasks are failing:

> @aws-cdk/aws-stepfunctions-tasks: Error: Some stacks have changed. To verify that they still deploy successfully, run: 'yarn integ eks/integ.call.js'

**Comment by ryparker** [comment]
> Looks good to me, but the integration tests of aws-stepfunctions-tasks are failing:
> 
> > @aws-cdk/aws-stepfunctions-tasks: Error: Some stacks have changed. To verify that they still deploy successfully, run: 'yarn integ eks/integ.call.js'

Done 👍🏻

**Review by otaviomacedo — APPROVED** [review]

**Comment by ryparker** [comment]
A bug was introduced by [this PR](https://github.com/aws/aws-cdk/commit/cf222806f781c3476dd942c57787ad0f4924dc04) that broke the `onEvent` EKS cluster handler lambda. ESBuild was inlining the node_module `proxy-agent` which was trying to read a file that did not exist (because all dependencies were bundled into a single file).

e.g.
```ts
var contextify = fs.readFileSync('/var/task/contextify.js');
```

Error:
```log
ENOENT: no such file or directory, open '/var/task/contextify.js' Logs: /aws/lambda/test-fixed-nobundle-eks-wit-OnEventHandler42BEBAE0-s2cZwaWDW0xt at Object.openSync (fs.js:462:3) at Object.readFileSync (fs.js:364:35) at loadAndCompileScript (/var/task/index.js:29479:23) at ../aws-cdk/node_modules/vm2/lib/main.js (/var/task/index.js:29490:25) at __require (/var/task/index.js:26:44) at ../aws-cdk/node_modules/vm2/index.js (/var/task/index.js:30079:23) at __require (/var/task/index.js:26:44) at ../aws-cdk/node_modules/degenerator/dist/src/index.js (/var/task/index.js:30091:17) at __require (/var/task/index.js:26:44) at ../aws-cdk/node_modules/pac-resolver/dist/index.js (/var/task/index.js:30857:25) (RequestId: c44d1357-fbce-4f96-8c23-b865c2c3aaff)
```

I created a [revert of this PR](https://github.com/aws/aws-cdk/pull/16651) and a [fresh PR with the corrected fix](https://github.com/aws/aws-cdk/pull/16652).

The updated fix:
- Replaces [`proxy-agent`](https://github.com/TooTallNate/node-proxy-agent) with the smaller module [`http-proxy-agent`](https://github.com/TooTallNate/node-http-proxy-agent) (reduced bundle size from 1.5mb -> 18.5kb)
- Adds configuration for the `NodeJsFunction`'s `bundling` param to not inline `http-proxy-agent`.
- Adds configuration for the `NodeJsFunction`'s `bundling` param to exclude `aws-sdk` from the bundle.
- Adds `esbuild` as a dev dependency of the `aws-eks` package (it is required by `NodeJsFunction` for tests).

**Pull request merged**